apps/src/code-studio/components/libraries/LibraryCreationDialog.jsx

```diff
@@ -103,113 +78,112 @@ class LibraryCreationDialog extends React.Component {
-      let functionsList = libraryParser.getFunctions(response.source);
-      if (!functionsList || functionsList.length === 0) {
-        this.setState({publishState: PublishState.NO_FUNCTIONS});
-        return;
-      }
-      let librarySource = response.source;
-      if (response.libraries) {
-        response.libraries.forEach(library => {
-          librarySource =
-            libraryParser.createLibraryClosure(library) + librarySource;
-        });
-      }
-      this.setState({
-        libraryName: libraryParser.suggestName(
-          dashboard.project.getLevelName()
-        ),
-        librarySource: librarySource,
-        publishState: PublishState.DONE_LOADING,
-        sourceFunctionList: functionsList
-      });
-    });
+    loadLibrary(
+      () => this.setState({dialogState: DialogState.CODE_ERROR}),
+      () => this.setState({dialogState: DialogState.NO_FUNCTIONS}),
+      libraryDetails =>
+        this.setState({
+          dialogState: DialogState.DONE_LOADING,
+          libraryDetails: libraryDetails
+        })
+    );
   };
 
   handleClose = () => {
-    this.setState({publishState: PublishState.LOADING});
+    this.setState({dialogState: DialogState.LOADING});
     this.props.onClose();
   };
 
-  copyChannelId = () => {
-    this.channelId.select();
-    document.execCommand('copy');
+  displayContent = () => {
+    const {libraryClientApi} = this.props;
+    const {libraryDetails} = this.state;
+    return (
+      <LibraryPublisher
+        onPublishSuccess={libraryName =>
+          this.setState({
+            dialogState: DialogState.PUBLISHED,
+            libraryName: libraryName
+          })
+        }
+        libraryDetails={libraryDetails}
+        libraryClientApi={libraryClientApi}
+      />
+    );
   };
 
-  publish = () => {
-    let {
-      librarySource,
-      libraryName,
-      libraryDescription,
-      selectedFunctions,
-      sourceFunctionList
-    } = this.state;
-    let functionsToPublish = sourceFunctionList.filter(sourceFunction => {
-      return selectedFunctions[sourceFunction.functionName];
-    });
-
-    if (!(libraryDescription && functionsToPublish.length > 0)) {
-      this.setState({publishState: PublishState.INVALID_INPUT});
-      return;
+  render() {
+    let bodyContent;
+    const {dialogState, libraryName} = this.state;
+    const {dialogIsOpen} = this.props;
+    switch (dialogState) {
+      case DialogState.LOADING:
+        bodyContent = <LoadingDisplay />;
+        break;
+      case DialogState.PUBLISHED:
+        bodyContent = <SuccessDisplay libraryName={libraryName} />;
+        break;
+      case DialogState.CODE_ERROR:
+        bodyContent = <ErrorDisplay message={i18n.libraryCodeError()} />;
+        break;
+      case DialogState.NO_FUNCTIONS:
+        bodyContent = <ErrorDisplay message={i18n.libraryNoFunctionsError()} />;
+        break;
+      default:
+        bodyContent = this.displayContent();
```

Comment:
It seems kind of weird to be using the default case as a legitimate/common case. Would it make sense to add a `DialogState` to cover this case (so that the options of DialogState actually exhaustively cover the cases, and the default case of this switch statement would catch if we like, add a new DialogState or something like that?)